doc/development/powershell.md

```diff
@@ -0,0 +1,179 @@
+
+## Style
+
+Powershell is a very flexible language. In engsys we try to use a smaller subset of what's possible to keep programs consistent and their behavior predictable.
+
+- Declare types for function parameters (e.g. `function foo([string]$bar) {}`.
+- Add formatted documentation for entrypoints like top-level scripts or publicly exposed functions.
+  ```
+  <#
+  .SYNOPSIS
+  <blah>
+  
+  .DESCRIPTION
+  <blah>
+  
+  .PARAMETER baz
+  <blah>
+  #>
+  function foobar([string]$baz)
+  ```
+- Avoid complex pipeline statements.
+- Put curly braces `{` for opening function blocks on a new line.
+    ```
+    function foo()
+    {
+    }
+    ```
+- Use explicit `return` statements for returning function/script block values.
+- Favor simple variable substitution in strings (see [variable substitution docs](https://docs.microsoft.com/en-us/powershell/scripting/learn/deep-dives/everything-about-string-substitutions?view=powershell-7.2)).
+  - For simple variables in a string, use syntax like `"my name is $username"`. Do not use `"my name is $($username)"`.
+  - For properties, use syntax like `"my name is $($user.name)"`.
+  - Only use curly braces when the delineation is unclear, e.g. `"Updating all ${itemType}s"`
+- Use full CmdLet names for better scripts.
+  - Favor `$foobar | Where-Object {} | ForEach-Object {}` over `$foobar | ? {} | % {}`.
+  - Read https://devblogs.microsoft.com/scripting/best-practice-for-using-aliases-in-powershell-scripts/ for some general reasons to avoid using aliases in scripts but even sense then we now have more reasons because of the cross-platform differences that pwsh brings. The more we depend on powershell scripts we need to be aware of ways to make them work on all our OS's. Of course we are using pwsh which runs on all OS's but default alias on the different OS's may change. [Two examples we hit](https://github.com/Azure/azure-sdk-tools/commit/0301c21a37de6e429f80cdb78a5f0adf41fafd12) while trying to run scripts on linux and windows:
+    - `mkdir` - On Windows this is an alias on top of New-Item and will happily create any missing directories in the directory hierarchy, while on linux this is the shell command and will need to be passed `-p` in order to create the missing directories. In order to make this work consistently use the cmdlet on all OS's call `New-Item -ItemType Directory` instead. 
+    - `echo` - This one isn't related to powershell it is more related to the script task where on windows echo cannot use quotes but on linux it requires quotes around the string. It is just another example where we need to be careful when writing steps we expect to work across OS's and in this case it might be better to use pwsh given we can make it consistent across OS's.
+- Analyze your code via an editor extension or on the command line with `Invoke-ScriptAnalyzer <script path>`.
```

Comment:
We should also add some tips with pointers to some of our helper for things like logging and retry. In particular we probably should talk about the different type of logging.

Reply:
I'll add some links to the logging. For retry, I can link to that once I add the common handler.